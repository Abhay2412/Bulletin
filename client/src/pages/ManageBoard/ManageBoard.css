a {
    text-decoration-color:  rgba(104, 57, 13, 0.54);
}

.manageHeaderBox{

    display: flex;
    margin: auto;
    padding: 10px;
    margin: auto;
    justify-content: center;

}

.managePhotoBox {
    width: 25%;
    justify-content: center;
    text-align: center;
    align-items: center;
}

.profileCircle {
    min-width: 100px;
    width: 25%;
}

.underlineButton {
    padding-top: 1vh;
    text-decoration: underline;
}

.boardNameBox {
    display: flex;
    flex-direction: column;
    max-width: fit-content;
    justify-content: center;
    padding-top: 2%;
}

.duplicateBoardBox {
    width: 20%;
    display: flex;
    flex-direction: column;
    justify-content: center;
    margin-right: 10%;
}

.newMemberInvitationBox {
    display: flex;
    justify-content: center;
    align-items: center;
    padding-top: 5px;
    padding-bottom: 5px;
}

.memberInfoBox {
    padding-left: 10vw;
    padding-right: 10vw;
    padding-bottom: 5vh;
}

.memberInfoHeaderBox {
    text-align: left;
    display: inline;
}

.permissionsGrid {
    display: flex;
    flex-direction: row;
    align-items: center;
}

.leaveBoardButtons {
    display: flex;
    flex-direction: row;
    justify-content: space-between;
    padding-left: 20px;
    padding-right: 20px;
    padding-bottom: 30px;
    padding-top: 20px;
}
.overlay{
    z-index: 2;
}
.InputText{
    width: 50%;
}
.boardDescription textarea {
    margin-top: 10px;
    height: 250px;
    width: 100%;
    min-height: 100%;
    resize: none;
}