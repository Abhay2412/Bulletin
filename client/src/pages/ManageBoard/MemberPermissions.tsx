import { Box, Checkbox, Typography } from "@mui/material";

interface Props {
  name: string;
  viewOn: boolean;
  editOn: boolean;
}

//Member Permissions Checkboxes on Manage Board - Title + View + Edit
export default function MemberPermissions({ name, viewOn, editOn }: Props) {
  return (
    <>

      <Box height={"60px"} textAlign="center">
        <Typography color={"primary"} variant="subtitle1" fontWeight={"bold"}>
          {name}
        </Typography>
      </Box>

      <Box
        height="50px"
        display={"flex"}
        flexDirection="row"
        alignItems={"center"}
        justifyContent="center"
      >
        <Checkbox defaultChecked={viewOn} />
      </Box>

      <Box
        height="50px"
        display={"flex"}
        flexDirection="row"
        alignItems={"center"}
        justifyContent="center"
      >
        <Checkbox defaultChecked={editOn} />
      </Box>
    </>
  );
}
