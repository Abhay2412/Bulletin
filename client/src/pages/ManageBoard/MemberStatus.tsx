import React from "react";
import { Box } from "@mui/system";
import {
  Typography,
  Button,
  Select,
  MenuItem,
  SelectChangeEvent,
} from "@mui/material";

interface Props {
  name: string;
  email: string;
  role: string;
}

//Member status display on manage board - name, email, role and remove
export default function MemberStatus({ name, email, role }: Props) {
  const [newRole, setNewRole] = React.useState("");

  const handleChange = (event: SelectChangeEvent) => {
    setNewRole(event.target.value as string);
  };

  return (
    <Box display={"flex"} flexDirection={"row"} justifyContent="space-evenly">
      <Box
        width="30%"
        sx={{ border: "1px solid rgb(104, 57, 13)" }}
        alignItems="center"
        display={"flex"}
        justifyContent="center"
      >
        <Typography color={"primary"}>{name}</Typography>
      </Box>

      <Box
        width="40%"
        sx={{ border: "1px solid rgb(104, 57, 13)" }}
        alignItems="center"
        display={"flex"}
        justifyContent="center"
      >
        <Typography color={"primary"}>{email}</Typography>
      </Box>

      <Box
        width="25%"
        sx={{ border: "1px solid rgb(104, 57, 13)" }}
        alignItems="center"
        display={"flex"}
        justifyContent="center"
      >
        {/* <Typography color={"primary"}> */}
        <Select
          sx={{ width: "100%" }}
          // labelId="demo-simple-select-label"
          // id="demo-simple-select"
          variant="outlined"
          // value={role}
          color="primary"
          defaultValue={role}
          onChange={handleChange}
        >
          <MenuItem color="primary" value={"Admin"}>
            Admin
          </MenuItem>
          <MenuItem color='primary' value={"Member"}>Member</MenuItem>
        </Select>
        {/* {role} */}
        {/* </Typography> */}
      </Box>

      <Box
        width="5%"
        display={"flex"}
        flexDirection="row"
        alignItems={"center"}
      >
        <Button>
          <Typography fontWeight={"bold"} color="primary" variant="h6">
            X
          </Typography>
        </Button>
      </Box>
    </Box>
  );
}
