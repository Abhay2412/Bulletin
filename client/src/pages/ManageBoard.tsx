import { Typography, Container, Box, Button } from "@mui/material";
import TextField from "@mui/material/TextField";
import MemberPermissions from "../components/MemberPermissions";
import MemberStatus from "../components/MemberStatus";
import avatar from "../imgs/avatar.jpg";

interface Props {
  name: string;
}

export default function ManageBoard({ name }: Props) {
  const mockMemberInfo = [
    { id: 1, name: "Liane Doe", email: "liane.doe@gmail.com", role: "Admin" },
    { id: 2, name: "Dad Doe", email: "dad.doe@gmail.com", role: "Admin" },
    { id: 3, name: "Logan Doe", email: "logan.doe@gmail.com", role: "Member" },
    { id: 4, name: "Aly Doe", email: "aly.doe@gmail.com", role: "Member" },
  ];

  const permissionHeaders = [
    { id: 1, name: "Notes", viewOn: true, editOn: false },
    { id: 2, name: "Lists", viewOn: true, editOn: true },
    { id: 3, name: "Expenses", viewOn: false, editOn: false },
    { id: 4, name: "Calendar", viewOn: true, editOn: false },
    { id: 5, name: "Personal Reminders", viewOn: true, editOn: true },
  ];

  return (
    <Container maxWidth={false} sx={{}}>
      {/* Change to type */}
      <a href="/" style={{ textDecorationColor: "rgba(104, 57, 13, 0.54)" }}>
        <Typography
          variant="subtitle1"
          color="primary.light"
          align="left"
          sx={{ pt: "1vh" }}
        >
          Back to {name} Board - Main
        </Typography>
      </a>

      <Box sx={{ pt: "5vh" }}>
        <Box
          sx={{ maxHeight: "8%" }}
          justifyItems="left"
          textAlign="left"
          display="flex"
          justifyContent="start"
        >
          <Box
            width={"25%"}
            justifyContent="center"
            textAlign="center"
            alignItems="center"
            sx={{ pr: "10%" }}
          >
            {/* For the board image */}
            <img
              src={avatar}
              style={{ borderRadius: "100%", minWidth: "100px" }}
              width="25%"
            ></img>
            <Button
              sx={{ display: "block", alignItems: "center", margin: "auto" }}
            >
              <Typography
                variant="subtitle1"
                color="primary.light"
                sx={{ pt: "1vh", textDecoration: "underline" }}
              >
                Change Board Picture
              </Typography>
            </Button>
          </Box>
          <Box
            display="flex"
            flexDirection="column"
            width="45%"
            justifyContent={"center"}
            paddingTop="2%"
          >
            {/* For the Board name */}
            <Typography variant="h5" color="primary" fontWeight="bold">
              {name}
            </Typography>
            <Button
              disableElevation={true}
              disableFocusRipple={true}
              disableRipple={true}
              disableTouchRipple={true}
              sx={{
                display: "block",
                margin: "auto",
                justifyContent: "left",
                marginLeft: "0",
                marginTop: "0",
              }}
            >
              <Typography
                variant="subtitle1"
                color="primary.light"
                align="left"
                sx={{ pt: "1vh", textDecoration: "underline" }}
              >
                Rename Board
              </Typography>
            </Button>
          </Box>
          <Box
            width={"20%"}
            display="flex"
            flexDirection={"column"}
            justifyContent="center"
            marginRight={"10%"}
          >
            <Button
              variant="contained"
              color="secondary"
              sx={{ borderRadius: "50px", border: "1px solid #68390D" }}
            >
              <Typography color="primary" variant="h6">
                Duplicate Board
              </Typography>
            </Button>
          </Box>
        </Box>
      </Box>
      <hr style={{ color: "rgb(104, 57, 13, 0.2)", borderWidth: "0.5px" }}></hr>

      <Box
        display={"flex"}
        flexDirection={"row"}
        justifyContent="center"
        alignItems={"center"}
        paddingTop="5px"
        paddingBottom={"5px"}
      >
        <Box width={"20%"}>
          <Typography color="primary" variant="h6">
            Invite New Member{" "}
          </Typography>
        </Box>
        <Box width={"30%"}>
          <TextField
            variant="outlined"
            color="primary"
            label="Email Address"
            sx={{ width: "90%" }}
          />
        </Box>

        <Box width={"20%"}>
          <Button
            variant="contained"
            color="secondary"
            sx={{ borderRadius: "50px", border: "1px solid #68390D" }}
          >
            <Typography color="primary" variant="h6">
              Send Invite{" "}
            </Typography>
          </Button>
        </Box>
      </Box>

      <hr style={{ color: "rgb(104, 57, 13, 0.2)", borderWidth: "0.5px" }}></hr>

      <Box paddingLeft={"10%"} paddingRight={"10%"} paddingBottom="20px">
        <Typography color="primary" variant="h6" textAlign={"left"}>
          Members
        </Typography>

        <Box display={"flex"} flexDirection="column">
          <Box
            display="flex"
            flexDirection="row"
            justifyContent={"space-between"}
          >
            <Box textAlign={"left"} paddingLeft="20px" width="30%">
              <Typography
                color="primary"
                variant="subtitle2"
                textAlign={"left"}
                display="inline"
              >
                Name
              </Typography>
            </Box>

            <Box textAlign={"left"} paddingLeft="20px" width="40%">
              <Typography
                color="primary"
                variant="subtitle2"
                textAlign={"left"}
                display="inline"
              >
                Email
              </Typography>
            </Box>

            <Box textAlign={"left"} paddingLeft="20px" width="25%">
              <Typography
                color="primary"
                variant="subtitle2"
                textAlign={"left"}
                display="inline"
              >
                Role
              </Typography>
            </Box>

            <Box width="5%"></Box>
          </Box>

          {mockMemberInfo.map((members) => {
            return (
              <MemberStatus
                name={members.name}
                email={members.email}
                role={members.role}
              ></MemberStatus>
            );
          })}
        </Box>
      </Box>

      <hr style={{ color: "rgb(104, 57, 13, 0.2)", borderWidth: "0.5px" }}></hr>

      <Box paddingLeft={"10%"} paddingRight={"10%"} paddingBottom="20px">
        <Typography color="primary" variant="h6" textAlign={"left"}>
          Member Permissions
        </Typography>

        <Box
          display={"flex"}
          flexDirection="row"
          justifyContent={"space-evenly"}
          paddingTop={"20px"}
        >
          <Box>
            <Box
              display="flex"
              flexDirection={"row"}
              alignItems="center"
              height="60px"
            ></Box>

            <Box
              height="50px"
              justifyContent={"center"}
              display="flex"
              flexDirection={"row"}
              alignItems="center"
            >
              <Typography
                color="primary"
                variant="subtitle1"
                fontWeight="bold"
                textAlign={"center"}
              >
                View
              </Typography>
            </Box>

            <Box
              height="50px"
              display="flex"
              flexDirection={"row"}
              alignItems="center"
            >
              <Typography
                color="primary"
                variant="subtitle1"
                fontWeight="bold"
                textAlign={"center"}
              >
                Edit
              </Typography>
            </Box>
          </Box>

          {permissionHeaders.map((permissions) => {
            return (
              <Box>
                <MemberPermissions
                  name={permissions.name}
                  viewOn={permissions.viewOn}
                  editOn={permissions.editOn}
                />
              </Box>
            );
          })}
        </Box>
      </Box>

      <hr style={{ color: "rgb(104, 57, 13, 0.2)", borderWidth: "0.5px" }}></hr>

      <Box
        display={"flex"}
        flexDirection="row"
        justifyContent={"space-between"}
        paddingLeft="20px"
        paddingRight={"20px"}
        paddingBottom="30px"
        paddingTop={"20px"}
      >
        <Button
          variant="contained"
          color="secondary"
          sx={{
            width: "20%",
            borderRadius: "50px",
            border: "1px solid #68390D",
          }}
        >
          <Typography color="primary" variant="h6">
            Leave Board{" "}
          </Typography>
        </Button>

        <Button
          variant="contained"
          color="warning"
          sx={{
            width: "20%",
            borderRadius: "50px",
            border: "1px solid #68390D",
          }}
        >
          <Typography color="primary" variant="h6">
            Delete Board{" "}
          </Typography>
        </Button>
      </Box>
    </Container>
  );
}
