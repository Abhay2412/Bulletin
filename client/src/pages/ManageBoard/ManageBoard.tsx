import React, { useEffect, useState } from "react";
import { Typography, Container, Box, Button, Modal } from "@mui/material";
import TextField from "@mui/material/TextField";
import MemberStatus from "./MemberStatus";
import Avatar from "@mui/material/Avatar";
import "./ManageBoard.css";
import { useParams } from "react-router-dom";
import axiosInstance from "../../axios";
import { Snackbar } from "@material-ui/core";
import { Alert, AlertColor } from "@mui/material";
import { useNavigate } from "react-router-dom";
import { useTypedSelector } from "../../hooks/ReduxHooks";
import { selectUserData } from "../../actions/UserActions/UserSelector";
import { Board, Member, User } from "../../../../types";

interface CreateBoardErrors {
  name: string;
  description: string;
}

export default function ManageBoard() {
  const userData = useTypedSelector(selectUserData);
  const navigate = useNavigate();
  const NAME_LIMIT = 20;
  const DESCRIPTION_LIMIT = 50;
  const params = useParams();
  const [popupState, setPopupState] = useState(false);
  const [editBoard, setEditBoard] = useState(false);
  const [board, setBoard] = useState<Board>({
    name: "",
    description: "",
    budgets: [],
    events: [],
    expenses: [],
    id: null,
    lists: [],
    members: [],
    notes: [],
    permissions: {
      editCalendar: true,
      editExpenses: true,
      editLists: true,
      editNotes: true,
      editPersonalReminders: true,
      viewCalendar: true,
      viewExpenses: true,
      viewLists: true,
      viewNotes: true,
      viewPersonalReminders: true,
    },
    tags: [],
  });
  const [members, setMembers] = useState<Array<Member>>([]);
  const [users, setUsers] = useState<Array<User>>([]);
  const [errors, setErrors] = useState<CreateBoardErrors>({
    name: "",
    description: "",
  });
  const [invitationEmail, setInvitationEmail] = useState<string>("");
  const [message, setMessage] = useState("");
  const [messageOpen, setMessageOpen] = useState(false);
  const [messageSeverity, setMessageSeverity] = useState<AlertColor>("success");

  useEffect(() => {
    let success = true;
    axiosInstance
      .get("/readBoard", { params: { boardID: params.board_id } })
      .then((res) => {
        console.log(res);
        if (success) {
          setBoard(res.data.content);
        }
      })
      .catch((err) => {
        console.log("error getting board: ", err);
        success = false;
      });
    axiosInstance
      .get("/readMembers", { params: { boardID: params.board_id } })
      .then((res) => {
        console.log(res);
        if (success) {
          setMembers(res.data.content);
        }
      })
      .catch((err) => {
        console.log("error getting users: ", err);
        success = false;
      });
  }, [editBoard]);

  useEffect(() => {
    setUsers([]);
    members.forEach((member) => {
      axiosInstance
        .get("/readUser", { params: { userID: member.userID } })
        .then((res) => {
          console.log(res);
          if (res.data.status === "success") {
            setUsers((prev) => [...prev, res.data.content]);
          }
        })
        .catch((err) => {
          console.log("error getting users: ", err);
        });
    });
  }, [members]);

  const handleChange =
    (prop: keyof Board) => (event: React.ChangeEvent<HTMLTextAreaElement>) => {
      setBoard({ ...board, [prop]: event.target.value });
    };

  const handleInvitationEmailChange = (event: any) => {
    setInvitationEmail(event.target.value);
  };

  const validateData = () => {
    errors.name = "";
    errors.description = "";

    let errorsExits = false;

    if (!board.name || board.name.length > NAME_LIMIT) {
      errors.name = "Please enter a valid board Name";
      errorsExits = true;
    }

    if (!board.description || board.description.length > DESCRIPTION_LIMIT) {
      errors.description = "Please enter a valid board description.";
      errorsExits = true;
    }

    setErrors({ ...errors });
    return !errorsExits;
  };

  const handleSaveBoard = () => {
    if (!validateData()) {
      return;
    }

    let success = true;

    axiosInstance
      .get("/readBoard", { params: { boardID: params.board_id } })
      .then((boardRes) => {
        axiosInstance
          .put(
            "/updateBoard",
            {
              ...boardRes.data.content,
              name: board.name,
              description: board.description,
            },
            { params: { boardID: params.board_id } }
          )
          .then((res) => {
            if (success) {
              setMessage("Board is Edited");
              setMessageOpen(true);
              setMessageSeverity("success");
              setPopupState(false);
            } else {
              setMessage("board cannot be edited.");
              setMessageSeverity("error");
              setMessageOpen(true);
            }
          })
          .catch((err) => {
            console.log("error editing board: ", err);
            success = false;
          });
      });
  };

  const deleteMember = (userBoardid: string, id: string, isLeave: boolean) => {
    const admins = members.filter((member) => member.role === "admin");
    console.log("admins is: ", admins);
    if (admins.length == 1 && admins[0].userID === id) {
      setMessage(
        "The only admin cannot be removed. Please assign a new admin first."
      );
      setMessageSeverity("error");
      setMessageOpen(true);
      return;
    }
    console.log(id);
    axiosInstance
      .delete("/deleteMember", {
        params: { boardID: params.board_id, memberID: id },
      })
      .then((res) => {
        console.log("user is deleted: " + res);
        // axiosInstance
        //   .get("./getUser", { params: { user_id: id } })
        //   .then((res) => {
        //     console.log(res);
        //     const removedUserData = res.data.user;
        //     const newBoards = res.data.user.boards.filter(
        //       (board: any) => board !== params.board_id
        //     );
        //     console.log(newBoards);
        //     const newRemovedUserData = {
        //       ...removedUserData,
        //       boards: newBoards,
        //     };
        //     axiosInstance
        //       .put("./editUser", newRemovedUserData)
        //       .then(() => {
        if (isLeave) {
          navigate("/boardsView");
        }
        //         console.log("new removed user data is: ", newRemovedUserData);
        //         console.log("user is edited");
        //         const index = members.findIndex(
        //           (x: Member) => x.userID === newRemovedUserData.id
        //         );
        //         if (index > -1) {
        //           members.splice(index, 1);
        //           setMembers([...members]);
        //         }
        //       })
        //       .catch((err) => {
        //         console.log("error editing user data: ", err);
        //       });
        //   })
        //   .catch((err) => {
        //     console.log("error editing user data: ", err);
        //   });

        setMessage("User is removed");

        setMessageSeverity("success");
        setMessageOpen(true);
      });
  };

  const sendInvitation = () => {
    axiosInstance
      .get("/readUserByEmail", { params: { email: invitationEmail } })
      .then(async (res) => {
        if (
          res.data.status === "success" &&
          members.find((mem) => mem.userID === res.data.content.id)
        ) {
          // if (res.data.content.email) {
          setMessage("User already has access to the board");
          setMessageOpen(true);
          setMessageSeverity("warning");
          return;
        }

        console.log(res);
        const invitedUserData = res.data.content;
        // const Boards = [...invitedUserData.boards];
        // Boards.push(params.board_id);
        // const newInvitedUserData = { ...invitedUserData, boards: [...Boards] };
        // axiosInstance
        //   .put("./editUser", newInvitedUserData)
        //   .then(() => {
        //     console.log("user is edited");
        //   })
        //   .catch((err) => {
        //     console.log("error editing user data: ", err);
        //   });

        const newMember: Member = {
          id: null,
          role: "member",
          userID: invitedUserData.id,
        };

        axiosInstance
          .post("/createMember", newMember, {
            params: { boardID: params.board_id },
          })
          .then((createMemberRes) => {
            console.log(createMemberRes);
            console.log("user added to board");
            setEditBoard(false);
            setInvitationEmail("");
            setMessage("User has been added to Board");
            setMessageOpen(true);
            setMessageSeverity("success");
          })
          .catch((err) => {
            console.log("error adding user to board: ", err);
          });
      })
      .catch((err) => {
        console.log("User Not Found", err);
        setMessage("User Not Found");
        setMessageOpen(true);
        setMessageSeverity("error");
      });
  };

  const handleLeaveButton = () => {
    const index = members.findIndex((x: Member) => x.userID === userData.id);
    if (index > -1) {
      deleteMember(params.board_id || "", members[index].userID || "", true);
    }
  };

  return (
    <Container maxWidth={false}>
      {popupState ? (
        <Modal
          open={true}
          onClose={() => setPopupState(false)}
          aria-labelledby="modal-modal-title"
          aria-describedby="modal-modal-description"
        >
          <div className="overlayBox">
            <div className="noteHeader">
              <Typography variant="h6">Edit Board</Typography>
              <Button onClick={() => setPopupState(false)}>
                <Typography variant="h5">X</Typography>
              </Button>
            </div>
            <Typography
              className="inputTitle"
              variant="h6"
              color="primary"
              sx={{ fontWeight: "bold" }}
            >
              Board Name
            </Typography>
            <TextField
              className="InputText"
              variant="outlined"
              value={board.name}
              onChange={handleChange("name")}
              focused
              id="name-text"
              inputProps={{
                maxLength: NAME_LIMIT,
              }}
              helperText={`${board.name.length}/${NAME_LIMIT}  ${errors.name}`}
              error={errors.name !== ""}
            />
            <Typography
              className="inputTitle"
              variant="h6"
              color="primary"
              sx={{ fontWeight: "bold" }}
            >
              Description
            </Typography>
            <TextField
              className="InputText"
              variant="outlined"
              value={board.description}
              onChange={handleChange("description")}
              focused
              align-items="left"
              id="description-text"
              inputProps={{
                maxLength: DESCRIPTION_LIMIT,
              }}
              rows="3"
              helperText={`${board.description.length}/${DESCRIPTION_LIMIT}  ${errors.description}`}
              multiline
              style={{ fontWeight: "bold" }}
              error={errors.description !== ""}
            />
            <div className="saveDiv">
              <Button
                className="saveButton"
                onClick={handleSaveBoard}
                id="saveButton"
              >
                Save Board
              </Button>
            </div>
          </div>
        </Modal>
      ) : null}
      <a href={"/board/" + params.board_id}>
        <Typography
          variant="subtitle1"
          color="primary.light"
          align="left"
          sx={{ pt: "1vh" }}
        >
          Back to {board.name} Board - Main
        </Typography>
      </a>

      {/* General Board Info: Profile Picture, Name, Duplicate Board */}
      <Box sx={{ pt: "5vh" }}>
        <Box className="manageHeaderBox">
          <Box className="managePhotoBox">
            {/* For the board image */}
            <Avatar
              className="profileCircle"
              sx={{
                background: "#f0e6db",
                color: "#AA896B",
                fontWeight: "bold",
                width: " 80px",
                height: "80px",
                fontSize: "60px",
                marginTop: "10px",
              }}
            >
              {board.name.split(" ")[0][0]}
            </Avatar>
          </Box>
          <Box className="boardNameBox">
            {/* For the Board name */}
            <Typography variant="h5" color="primary" fontWeight="bold">
              {board.name}
            </Typography>
            <Button
              disableElevation={true}
              disableFocusRipple={true}
              disableRipple={true}
              disableTouchRipple={true}
              className="renameButton"
              id="renameButton"
              onClick={() => setPopupState(true)}
            >
              <Typography
                className="underlineButton"
                variant="subtitle1"
                color="primary.light"
                textAlign={"left"}
              >
                Edit Board
              </Typography>
            </Button>
          </Box>
        </Box>
      </Box>

      <hr style={{ color: "rgb(104, 57, 13, 0.2)", borderWidth: "0.5px" }}></hr>

      {/* New Member Invitation */}
      <Box className="newMemberInvitationBox">
        <Box width={"20%"}>
          <Typography color="primary" variant="h6">
            Invite New Member
          </Typography>
        </Box>
        <Box width={"30%"}>
          <TextField
            label="Email Address"
            value={invitationEmail}
            onChange={handleInvitationEmailChange}
            sx={{ width: "90%" }}
            id="invitation-email-text"
          />
        </Box>

        <Box width={"20%"}>
          <Button
            variant="contained"
            color="secondary"
            sx={{ borderRadius: "50px", border: "1px solid #68390D" }}
            onClick={sendInvitation}
            id="sendInviteButton"
          >
            <Typography color="primary" variant="h6">
              Send Invite
            </Typography>
          </Button>
        </Box>
      </Box>

      <hr style={{ color: "rgb(104, 57, 13, 0.2)", borderWidth: "0.5px" }}></hr>

      {/* Member Information */}
      <Box className="memberInfoBox" id="memberInfoBox">
        <Typography color="primary" variant="h6" textAlign={"left"}>
          Members
        </Typography>

        <Box display={"flex"} flexDirection="column">
          <Box
            display="flex"
            flexDirection="row"
            justifyContent={"space-between"}
          >
            <Box textAlign={"left"} paddingLeft="20px" width="30%">
              <Typography
                className="memberInfoHeaderBox"
                color="primary"
                variant="subtitle2"
              >
                Name
              </Typography>
            </Box>

            <Box textAlign={"left"} paddingLeft="20px" width="40%">
              <Typography
                className="memberInfoHeaderBox"
                color="primary"
                variant="subtitle2"
              >
                Email
              </Typography>
            </Box>

            <Box textAlign={"left"} paddingLeft="20px" width="25%">
              <Typography
                className="memberInfoHeaderBox"
                color="primary"
                variant="subtitle2"
              >
                Role
              </Typography>
            </Box>

            <Box width="5%"></Box>
          </Box>
          {members.map((member, index) => {
            if (users[index])
              return (
                <MemberStatus
                  onDelete={deleteMember}
                  userBoardid={params.board_id || ""}
                  id={member.id || ""}
                  name={users[index].name}
                  email={users[index].email}
                  role={member.role}
                />
              );
          })}
        </Box>
      </Box>

      <hr style={{ color: "rgb(104, 57, 13, 0.2)", borderWidth: "0.5px" }}></hr>
      <Box className="leaveBoardButtons">
        <Button
          variant="contained"
          color="secondary"
          sx={{
            width: "20%",
            borderRadius: "50px",
            border: "1px solid #68390D",
          }}
          onClick={handleLeaveButton}
        >
          <Typography color="primary" variant="h6">
            Leave Board
          </Typography>
        </Button>

        {/* <Button
          variant="contained"
          color="warning"
          sx={{
            width: "20%",
            borderRadius: "50px",
            border: "1px solid #68390D",
          }}
          onClick={deleteBoard}
        >
          <Typography color="primary" variant="h6">
            Delete Board
          </Typography>
        </Button> */}
      </Box>
      <Snackbar
        open={messageOpen}
        autoHideDuration={6000}
        onClose={() => setMessageOpen(false)}
      >
        <Alert severity={messageSeverity} sx={{ width: "100%" }}>
          {message}
        </Alert>
      </Snackbar>
    </Container>
  );
}
