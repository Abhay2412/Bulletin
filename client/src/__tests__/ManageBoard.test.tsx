import React, { useState } from "react";
import ManageBoard from "../pages/ManageBoard/ManageBoard";
import { configure, mount, ReactWrapper, shallow } from "enzyme";
import Adapter from "@wojtekmaj/enzyme-adapter-react-17";
import { createMemoryHistory } from "history";
import { render as reactRender } from "@testing-library/react";
import { store } from "../store";
import { Provider } from "react-redux";
import { BrowserRouter } from "react-router-dom";

configure({ adapter: new Adapter() });

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useParams: () => ({
    board_id: "g7HGuvfnGe3nhQgAF8qY",
  }),
}));

describe("Testing <ManageBoard /> Component", () => {
  it("should render Manage-Board page correctly", () => {
    const app = shallow(
      <Provider store={store}>
        <BrowserRouter>
          <ManageBoard />
        </BrowserRouter>
      </Provider>
    );
    expect(app.getElements()).toMatchSnapshot();
  });
  describe("Need open modal setup", () => {
    const wrapper: ReactWrapper = mount(
      <Provider store={store}>
        <BrowserRouter>
          <ManageBoard />
        </BrowserRouter>
      </Provider>
    );

    beforeEach(() => {
      wrapper.find("button").find("#renameButton").simulate("click");
    });

    it("should change board title value", () => {
      wrapper
        .find("input")
        .find("#name-text")
        .simulate("change", {
          target: { value: "Edit Board Name" },
        });
      expect(
        wrapper.find("input").find("#name-text").get(0).props.value
      ).toEqual("Edit Board Name");
    });

    it("should display board name error", () => {
      wrapper
        .find("input")
        .find("#name-text")
        .simulate("change", {
          target: { value: "" },
        });

      wrapper.find("button").find("#saveButton").simulate("click");
      expect(
        wrapper.find("p").find("#name-text-helper-text").text()
      ).toEqual("0/20  Please enter a valid board Name");
    });

    it("should change board description value", () => {
        wrapper
          .find("textarea")
          .find("#description-text")
          .simulate("change", {
            target: { value: "Edit Board description" },
          });
        expect(
          wrapper.find("textarea").find("#description-text").get(0).props.value
        ).toEqual("Edit Board description");
      });
  
      it("should display board description error", () => {
        wrapper
          .find("textarea")
          .find("#description-text")
          .simulate("change", {
            target: { value: "" },
          });
  
        wrapper.find("button").find("#saveButton").simulate("click");
        expect(
          wrapper.find("p").find("#description-text-helper-text").text()
        ).toEqual("0/50  Please enter a valid board description.");
      });

      it("should update board name", () => {
        wrapper
          .find("textarea")
          .find("#description-text")
          .simulate("change", {
            target: { value: "description test" },
          });
        wrapper
        .find("input")
        .find("#name-text")
        .simulate("change", {
        target: { value: "name test" },
        });
  
        wrapper.find("button").find("#saveButton").simulate("click");
        expect(
          wrapper.find("h5").at(0).text()
        ).toEqual("name test");
      });
  });
  describe("Inviation email", () => {
        const wrapper: ReactWrapper = mount(
        <Provider store={store}>
            <BrowserRouter>
            <ManageBoard />
            </BrowserRouter>
        </Provider>
        );
        it("should change invitaton email value", () => {
            wrapper
            .find("input")
            .find("#invitation-email-text")
            .simulate("change", {
                target: { value: "test@mail.com" },
            });
            expect(
            wrapper.find("input").find("#invitation-email-text").get(0).props.value
            ).toEqual("test@mail.com");
        });
    });
    describe("Member Info box", () => {
        const wrapper: ReactWrapper = mount(
        <Provider store={store}>
            <BrowserRouter>
            <ManageBoard />
            </BrowserRouter>
        </Provider>
        );
        
        it("Member box should have three headers/columns", () => {
            expect(wrapper.find("#memberInfoBox").length).toEqual(3);
        });
    });

})