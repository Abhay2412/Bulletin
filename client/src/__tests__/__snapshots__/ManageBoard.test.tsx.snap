// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing <ManageBoard /> Component should render Manage-Board page correctly 1`] = `
Array [
  <Provider
    store={
      Object {
        "@@observable": [Function],
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
  >
    <BrowserRouter>
      <ManageBoard />
    </BrowserRouter>
  </Provider>,
]
`;

exports[`Testing <ManageBoard /> Component should render Member Permissions correctly 1`] = `
Array [
  <Provider
    store={
      Object {
        "@@observable": [Function],
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
  >
    <BrowserRouter>
      <MemberPermissions
        editOn={true}
        name="Test User"
        viewOn={true}
      />
    </BrowserRouter>
  </Provider>,
]
`;
